File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.java
class Solution {
    public String solution(int[] numbers, String hand) {
        int[][] keyPad ={{1, 2, 3}, {4, 5, 6}, {7, 8, 9} , {-1, 0 ,-1}};
        int[] left = {3, 0};
        int[] right = {3, 2};
        String answer = "";
        for(int i = 0; i < numbers.length; i++) {
           for(int j = 0; j < keyPad.length; j++) {
               for(int k = 0; k < keyPad[j].length; k++) {
                   if(numbers[i] == keyPad[j][k]) {
                       if(numbers[i] == 1 || numbers[i] == 4 || numbers[i] == 7) {
                       answer+="L";
                       left[0] = j;
                       left[1] = k;
                   }
                       else if(numbers[i] == 3 || numbers[i] == 6 || numbers[i] == 9) {
                       answer += "R";
                       right[0] = j;
                       right[1] = k;
                   }
                        else {
                            int x1 = Math.abs(j - left[0]);
                            int x2 = Math.abs(j - right[0]);
                            int y1 = Math.abs(k - left[1]);
                            int y2 = Math.abs(k - right[1]);
                            if(x1 + y1 < x2 + y2) {
                                answer += "L";
                                left[0] = j;
                                left[1] = k;
                            }
                            else if(x1 + y1 == x2 + y2) {
                                if(hand.equals("right")) {
                                    answer+= "R";
                                    right[0] = j;
                                    right[1] = k;
                                }
                                else {
                                    answer += "L";
                                    left[0] = j;
                                    left[1] = k;
                                }
                            }
                            else {
                                answer += "R";
                                right[0] = j;
                                right[1] = k;
                            }
                    
                   }
                 }
                   
                   
                  
               }
           }
        }
        return answer;
    }
}

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.js
const phone = [[1, 2, 3],
                   [4, 5, 6],
                   [7, 8, 9],
                   [-1,0,-1]];
    

function findPosition(position) {
    let x = 0, y = 0;
    for(let i = 0; i < phone.length; i++) {
        for (let j = 0; j < phone[i].length; j++) {
            if(position === phone[i][j]) {
                y = i;
                x = j;
            }
        }
    }
    return [y, x];
}

function calDistance(lf,rf,position,hand) {
    const L =Math.abs(lf[0]-position[0]) + Math.abs(lf[1] -  position[1]);
    const R = Math.abs(rf[0]-position[0]) + Math.abs(rf[1] - position[1]);
    if (L < R) return 'L';
    else if (R < L) return 'R';
    else {
        if (hand === 'left') return 'L';
        else return 'R';
    }
}
function solution(numbers, hand) {
    var answer = '';
    const left = [1,4,7];
    const right = [3,6,9];
    let lf = [3, 0];
    let rf = [3, 2];
    for(let i = 0; i < numbers.length; i++) {
        const number = numbers[i];
        if(left.includes(number)) {
            answer += 'L';
            lf = findPosition(number);
        } else if(right.includes(number)) {
            answer += 'R';
            rf = findPosition(number);
        } else {
            let position = [];
            position = findPosition(number);
            const find = calDistance(lf,rf,position,hand);
            if (find === 'L') lf = position;
            else rf = position;
            answer += find;
        }
    }
    return answer;
}